### ISSUE TITLE
Checkout goes back to address step after applying a coupon that's > order amount

### ISSUE THREAD

**Issue by justindrees** [spree/spree#6673]
### Issue
- Checkout goes back to Address step after applying a coupon which is more than the total order amount.
### Expected behavior
- This could be improved in various ways. Instead of sending the user back to the address step after applying the coupon and forcing him/her to click through the entire checkout flow, it could:
- stay on the payment step, update the order summary, remove/update the payment methods and let the user finalize the purchase at this step.
- remove the payment step (this is what happens now) and advance the user to the confirm step or the final step (right now the delivery step becomes the final step) where the purchase can be finalized.
#### Admin steps
1. Login to admin interface.
2. Click "Promotions".
3. Click button "New Promotion".
4. Enter attributes:  name = "test-coupon", code = "coupon", starts_at = today
5. Click "Create"
6. In Actions box, use action type "Create whole-order adjustment" and click "Add"
7. In Calculator, select "Flat Rate" and click Update.
8. In the action which was just created, add an amount for the promotion which will be greater than the total order amount.
9. Click "Update" to finalize the creation of the promotion.
#### User steps
1. Login in as existing user
2. View and add item to cart which will have a total order amount (including shipping and tax) that is less than the coupon amount. So, after the coupon is applied, this will result in the total order amount being zero.
3. Go to shopping cart and click button "Checkout"
4. Address should be prefilled. Click "Save and Continue."
5. On Delivery, use default shipping method and click "Save and Continue"
6. On Payment, enter coupon code for Promotion with rule "Only one use per user." Click "Save and Continue".
   -> Coupon is applied and user is then brought back to Address step.

**Comment by kurapatijayaram** [comment]
Ya ... Even I am facing this issue. Whenever order got updated the state changes to 'address'.

**Comment by antoineayoub** [comment]
Hey @justindrees, @kurapatijayaram , I'm facing this issue too ! have you solved it ?

Thx

**Comment by justindrees** [comment]
@antoineayoub nope, didn't pursue this further after discovering the bug.

**Comment by antoineayoub** [comment]
@justindrees I have found why but I don't now how to solve it. 

It comes from the checkout flow in the order model. When the state is "payment" it check if the total is greater than 0.00 if not it goes back to address, the previous one.
``` ruby
  checkout_flow do
      go_to_state :address
      go_to_state :delivery
      go_to_state :payment, if: ->(order) { order.payment_required? }
      go_to_state :confirm, if: ->(order) { order.confirmation_required? }
      go_to_state :complete
      remove_transition from: :delivery, to: :confirm
    end

def payment_required?
      total.to_f > 0.0
end
```

**Comment by kushniryb** [comment]
This was fixed in 7b6feb284dfe43d24a78603c88b79be6dcfce17f
@bbonislawski We can close this.
